framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -957,6 +958,48 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
     }
 
     /**
+     * A helper method to check that the array type of actual varargs is a subtype of the
+     * corresponding required varargs, and issues "argument.invalid" error for it's not a subtype of
+     * the required one.
+     *
+     * <p>Note this method requires that type checking for the each elements of varargs is executed
+     * by caller.
+     *
+     * @see #checkArguments(List, List)
+     * @param invokedMethod the method type to be invoked
+     * @param tree method or constructor invocation tree
+     */
+    private void checkVarargs(AnnotatedExecutableType invokedMethod, Tree tree) {
+        if (!invokedMethod.isVarArgs()) {
+            return;
+        }
+
+        List<AnnotatedTypeMirror> formals = invokedMethod.getParameterTypes();
+        int numFormals = formals.size();
+        int lastArgIndex = numFormals - 1;
+        AnnotatedArrayType lastParamAnnotatedType = (AnnotatedArrayType) formals.get(lastArgIndex);
+        AnnotatedTypeMirror wrappedVarargsType = atypeFactory.getAnnotatedTypeVarargsArray(tree);
+
+        if (wrappedVarargsType == null) {
+            return;
+        }
+
+        // If the component type is intersection or including type variable by type inference,
+        // the comparison doesn't work well.
+        // We can consider that the both component type of actual and formal is Object,
+        // because type checking for elements will be done in checkArguments.
+        AnnotatedTypeMirror objectType =
+                atypeFactory.getAnnotatedType(elements.getTypeElement("java.lang.Object"));
+        lastParamAnnotatedType.setComponentType(objectType);
+        if (wrappedVarargsType instanceof AnnotatedArrayType) {
+            ((AnnotatedArrayType) wrappedVarargsType).setComponentType(objectType);
+        }
+
+        commonAssignmentCheck(
+                lastParamAnnotatedType, wrappedVarargsType, tree, "argument.type.incompatible");
```

Comment:
Should we use a distinct error key, e.g. `varargs.argument.type.incompatible`, to make the error message easier to understand?